### ISSUE TITLE
[prometheus] Unable to install in custom namespace

### ISSUE THREAD

**Issue by arpitjindal97** [prometheus-community/helm-charts#969]
**Describe the bug**
Trying to install prometheus in namespace named `tenant1`. I want to scrape resources present in this namespace only.

**Version of Helm and Kubernetes**:

Helm Version:

```console
$ helm version
version.BuildInfo{Version:"v3.5.4", GitCommit:"1b5edb69df3d3a08df77c9902dc17af864ff05d1", GitTreeState:"dirty", GoVersion:"go1.16.3"}
```

Kubernetes Version:

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:02:01Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Which chart**: prometheus

**Which version of the chart**: 13.8.0

**What happened**: helm threw error

**What you expected to happen**: helm should return success

**How to reproduce it**: Simply try to install chart on namespace `tenant1`

**Changed values of values.yaml** (only put values which differ from the defaults):

values.yaml

```console

kubeStateMetrics:
  enabled: false

alertmanager:
  enabled: false

rbac:
  create: true

server:
  useClusterRole: false
  namespaces:
    - tenant1
  global:
    external_labels:
      cluster: tenant1
  remoteWrite:
    - url: http://cortex.cortex/api/prom/push

namespace: tenant1
forceNamespace: tenant1
```

**The helm command that you execute and failing/misfunctioning**:

```console
helm upgrade -f values-prometheus.yml -n tenant1 prometheus prometheus-community/prometheus
```

**Anything else we need to know**: 

My use case is to deploy prometheus on namespace level. Prometheus should be able to discover services and pods based on annotation on own namespace. I tried to put `rbac.create: false` then helm succeeded but prometheus wasn't able to discover any service or pod on it's own. Logs of prometheus-server are below:

```
level=error ts=2021-05-13T13:45:14.612Z caller=klog.go:96 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.20.5/tools/cache/reflector.go:167: Failed to watch *v1.Service: failed to list *v1.Ser
vice: services is forbidden: User \"system:serviceaccount:tenant1:prometheus-server\" cannot list resource \"services\" in API group \"\" at the cluster scope"
level=error ts=2021-05-13T13:45:18.684Z caller=klog.go:96 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.20.5/tools/cache/reflector.go:167: Failed to watch *v1.Node: failed to list *v1.Node:
nodes is forbidden: User \"system:serviceaccount:tenant1:prometheus-server\" cannot list resource \"nodes\" in API group \"\" at the cluster scope"
level=error ts=2021-05-13T13:45:30.208Z caller=klog.go:96 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.20.5/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: po
ds is forbidden: User \"system:serviceaccount:tenant1:prometheus-server\" cannot list resource \"pods\" in API group \"\" at the cluster scope"
level=error ts=2021-05-13T13:45:36.862Z caller=klog.go:96 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.20.5/tools/cache/reflector.go:167: Failed to watch *v1.Endpoints: failed to list *v1.E
ndpoints: endpoints is forbidden: User \"system:serviceaccount:tenant1:prometheus-server\" cannot list resource \"endpoints\" in API group \"\" at the cluster scope"
```

If I put a static target in prometheus scape_config then it was scraping very well. So, the problem is it's not able to discover. I suspect it's related to `rbac` somehow. I'm unable to find a fix for it yet.

**Comment by danielhstahl** [comment]
I have a similar problem.  

```
kubeStateMetrics:
  enabled: false
alertmanager:
  useClusterRole: false
  # useExistingRole: false # set to name of role?
  # baseURL: "http://localhost:9093"

server:
  useExistingClusterRoleName: ${prometheusclusterrole}

  namespaces:
    - ${serviceplane}
    - ${sparkplane}
  
  baseURL: https://example.com/monitoring # see https://blog.cubieserver.de/2020/configure-prometheus-on-a-sub-path-behind-reverse-proxy/ and gateway/ingress.yml
  extraArgs:
    # blank since "monitoring" prefix will be stripped by nginx
    web.route-prefix: / # https://prometheus.io/docs/guides/tls-encryption/
```

I created an existing cluster role that is identical to the one automatically created when not providing an existing cluster role.  The two rolebindings associated with each namespace were created and appropriately tied to the service account.  But I still get 

`Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User \"system:serviceaccount:[namespace]:prometheus-server\" cannot list resource \"pods\" in API group \"\" at the cluster scope"`

**Comment by arpitjindal97** [comment]
@danielhstahl please also post yaml of clusterrole. I hope somebody from community might be able to point error.

**Comment by danielhstahl** [comment]
Its the same cluster role that is made automatically if you don't specificy cluster role.

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    meta.helm.sh/release-name: prometheus
    meta.helm.sh/release-namespace: ${namespace}
  labels:
    app: prometheus
  name: ${rolename}
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  - nodes/proxy
  - nodes/metrics
  - services
  - endpoints
  - pods
  - ingresses
  - configmaps
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - extensions
  - networking.k8s.io
  resources:
  - ingresses/status
  - ingresses
  verbs:
  - get
  - list
  - watch
- nonResourceURLs:
  - /metrics
  verbs:
  - get
```

**Comment by pgdagenais** [comment]
I have the same issue here using the same role than @danielhstahl is using.

I am able to create all the resources but prometheus is giving the same errors than you guys. I also tried using the `cluster-admin` role instead of my cluster role and I also have the same issue.

**Comment by pgdagenais** [comment]
I was able to identify the issue (I think). Since the default `kubernetes_sd_configs` is trying to get all the resources on a cluster scope level, you need to also modify the default rules that comes with the chart to match what you are trying to do.

**Comment by jacksgt** [comment]
Hey there! As you pointed out, this sounds very much related to the fact that by default Prometheus tries to scrape all Kubernetes namespaces.
`Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User \"system:serviceaccount:[namespace]:prometheus-server\" cannot list resource \"pods\" in API group \"\" at the cluster scope`
See also https://github.com/prometheus/prometheus/issues/2763#issuecomment-458054220

I have described the problem in more detail in [my blog post "Namespace-scoped Kubernetes Service Discovery for Prometheus"](https://blog.cubieserver.de/2021/namespace-scoped-kubernetes-service-discovery-for-prometheus/).
In summary, you should specify the following Helm values to tell Prometheus to scrape only a specific namespace:
```yaml
server:
  useExistingClusterRoleName: false
  namespaces:
    - "mynamespace"
 ```

Please report back if the issue still occurs afterwards.

**Comment by jeehyun-lee** [comment]
@jacksgt I tried `useExistingClusterRoleName: false` but it still having same error
```
level=error ts=2021-07-23T05:31:18.529Z caller=klog.go:96 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.20.5/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User \"system:serviceaccount:prometheus:tropicana-prometheus-server\" cannot list resource \"pods\" in API group \"\" at the cluster scope"
```

chart : prometheus-community/prometheus
chart version : 14.4.1

I'm curious about
1. there is no roleBinding (or clusterRoleBinding) when using `useExistingClusterRoleName: false`. Do I have to create myself?
2. I figure that server.namespace is only used by  creating roleBinding, only when useExistingClusterRoleName is not false but having some value. I read your blog post, but at last you said just using these helm values will work.
```
server:
  useExistingClusterRoleName: false
  namespaces:
    - "mynamespace"
```
How dose prometheus limit namespace with this?

**Comment by jacksgt** [comment]
@jeehyun-lee 
I assume you are using the default scrape_configs Helm value provided by this repository?
https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/values.yaml#L1339

Unfortunately those won't work, because they are missing the necessary namespace scoping, like this:
```
kubernetes_sd_configs:
  - job_name: 'foobar'
    namespaces:
      names:
        - "mynamespace"
```

There is also no dynamic way to inject those, so you will need to overwrite the entire scrape_config array (yes, that sucks). 

Is it more clear now?

**Comment by k-jingyang** [comment]
@jacksgt 

I had a similar issue and your solution helped. Thanks!  

Wondering if we can modify the Helm chart to inject the namespaces in when they are specified in `server.namespaces`?
